modules/loader.js

```diff
@@ -846,90 +842,87 @@ __private.snapshotFinished = err => {
+			async.waterfall(
+				[
+					waterfallCb => {
+						self.getRandomPeerFromNetwork((err, peer) => {
 							if (err) {
-								library.logger.error(err.toString());
+								failedAttemptsToLoad += 1;
 								library.logger.error(
-									`Failed to load blocks from: ${peer.string}`
+									'Failed to get random peer from network',
+									err.toString()
 								);
-								errorCount += 1;
+								return whilstCb();
 							}
-							loaded = lastValidBlock.id === lastBlock.id;
-							lastBlock = null;
-							lastValidBlock = null;
-							next();
+							__private.blocksToSync = peer.height;
+							const lastBlock = modules.blocks.lastBlock.get();
+							library.logger.info(
+								`Looking for common block with: ${peer.string}`
+							);
+							return waterfallCb(null, peer, lastBlock);
+						});
+					},
+					(peer, lastBlock, waterfallCb) => {
+						// No need to call getCommonBlock if height is Genesis block
+						if (lastBlock.height === 1) {
+							return waterfallCb(null, peer, lastBlock);
 						}
-					);
-				}
 
-				/**
-				 * Description of getCommonBlock.
-				 *
-				 * @todo Add @returns and @param tags
-				 * @todo Add description for the function
-				 */
-				function getCommonBlock(cb) {
-					library.logger.info(`Looking for common block with: ${peer.string}`);
-					modules.blocks.process.getCommonBlock(
-						peer,
-						lastBlock.height,
-						(err, commonBlock) => {
-							if (!commonBlock) {
+						modules.blocks.process.getCommonBlock(
+							peer,
+							lastBlock.height,
+							(err, commonBlock) => {
 								if (err) {
 									library.logger.error(err.toString());
+									failedAttemptsToLoad += 1;
+									return whilstCb();
 								}
-								library.logger.error(
-									`Failed to find common block with: ${peer.string}`
+								if (!commonBlock) {
+									failedAttemptsToLoad += 1;
+									library.logger.error(
+										`Failed to find common block with: ${peer.string}`
+									);
+									return whilstCb();
+								}
+								library.logger.info(
+									`Found common block: ${commonBlock.id} with: ${peer.string}`
 								);
-								errorCount += 1;
-								return next();
+								return waterfallCb(null, peer, lastBlock);
 							}
-							library.logger.info(
-								[
-									'Found common block:',
-									commonBlock.id,
-									'with:',
-									peer.string,
-								].join(' ')
-							);
-							return setImmediate(cb);
-						}
-					);
-				}
-				if (lastBlock.height === 1) {
-					loadBlocks();
-				} else {
-					getCommonBlock(loadBlocks);
-				}
-			});
+						);
+					},
+					(peer, lastBlock, waterfallCb) => {
+						modules.blocks.process.loadBlocksFromPeer(
+							peer,
+							(err, lastValidBlock) => {
+								if (err) {
+									failedAttemptsToLoad += 1;
+									library.logger.error(
+										`Failed to load blocks from: ${peer.string}`,
+										err.toString()
+									);
+									return whilstCb();
```

Comment:
This step is accepting `waterfallCb` and calling `whilstCb`, we should avoid this.

Reply:
How do you want to make it? The waterfall is inside the whilst. We need to stop the cycle from being propagated when we received an error.

Reply:
We can call `waterfallCb` with some error or just `true`, it will break the execution then.